### DISCUSSION TITLE
Book about NNI

### DISCUSSION THREAD

**Discussion by survexman** [microsoft/nni#4099] [Ideas]
Hi everyone!

For me, NNI is a very promising framework. I think it can be the best and very popular framework for AutoML and NAS tasks.

**There is an idea to write a book about NNI: "Hands-on Neural Network Intelligence".**

I'd like to give explanations of the main concepts and ideas. Give a lot of practical examples of hyper-parameter optimization and especially cover the NAS topic. 

I already have experience in writing, this is one of my books: 
https://www.amazon.com/Learning-Genetic-Algorithms-Python-capabilities-ebook/dp/B08WKBFHGV

I think that can help popularize the framework.

What do you think?

Any ideas and contributions appreciated :+1:

**Comment by scarlett2018** [comment]
@survexman - Appreciate for the comment and believes in NNI. I'm the PM for the project. We highly encourage our community contributors to share experience through examples, tutorials, books. 

Would you like to list out an outline for your idea? Feel free to use this thread to gather people joining as author(s). We can also help to distribute the recruit in other NNI fans channels.

**Reply by survexman** [reply]
@scarlett2018 , thank you for your reply!

At first, I want to say that I'll keep this book updated. It will not be too hard for me because I'm a permanent user of NNI. It means there'll be different editions that contain all actual updates.

This book is for practical data scientists and ML experts. This book does not cover the essentials of the machine and deep learning.

I want to choose PyTorch as a primary deep learning framework for practical examples in the book.

Here is the outline.

### **Chapter 1. Introduction to NN**

- Installation

- Trial, Search Space, Experiment
_Explaining and visualizing basic concepts of optimization in NNI context._

- Finding maxima of a multivariate function
_NNI can be used as an optimization engine for any problem. I want to start explaining NNI principles from the purely mathematical task of finding the maxima of a function using a simple hyper-parameter tuning engine._ 

- Interacting with NNI
_About NNI environment, its GUI, how to save results and etc._

### **Chapter 2. Hyper-parameter tuning**

_The Real-world time-series prediction LSTM model (weather: rain prediction) will be used as a practical example for this chapter._

- Preparing a model for hyper-parameter tuning

- Running Experiment

- Interpreting Results

- Debugging

### **Chapter 3. Hyper-parameter tuners**

_Description of main concepts of most popular tuners. The main idea is to give a reader an intuitive explanation of the most popular tuners. Because often developers do not understand which tuner to chose and how it works._

### **Chapter 4. Neural Architecture Search: Multi-trial**

_This is the most exciting part for me. I think NAS techniques sometimes are treated as very complicated methods. And usually, developers and data scientists use hyper-parameter tuning only. I'd like to help people go a little further here because it is not so hard as it seems._

_We will take the text classification problem as an example for this chapter._ 

- Constructing a search space for architecture search

- Running architecture search
_(with random strategy)_

- Exploration strategies (short overview of Random Strategy, RL Strategy, etc.)

- Comparing exploration strategies

### **Chapter 5. Neural Architecture Search: One-shot**

_Here I will describe the main idea of one-shot NAS and its practical implementations._

- What is one-shot NAS?

- ENAS

- DARTS

### **Chapter 6. Model Compression**

- What is model compression?
_Here I'll explain why model compression is important and why NAS algorithms usually do not support model compression._

- Trying to compress your model
_Simple practical example here._

- Pruning
_Introduction to pruners._

- Quantization
_Explanation and visualization of quantization._

### **Chapter 7. Advanced NNI techniques**

_Here we will cover advanced techniques of NNI usage. Distributed experiments, different platforms, speed up methods, etc._

Let me know what do you think?

**Comment by survexman** [comment]
@scarlett2018  - I can send a request to some publishers by myself, or maybe you can help me with this somehow?

**Reply by scarlett2018** [reply]
> @scarlett2018 - I can send a request to some publishers by myself, or maybe you can help me with this somehow?

Hi @survexman, there do have publishers contact me, let me connect some of them to you.

**Comment by survexman** [comment]
Happy to announce that the book about NNI has been published recently!

https://www.amazon.com/Automated-Learning-Neural-Network-Intelligence/dp/1484281489/

**Reply by shabnam-ataee** [reply]
Thanks! It is very useful, but I am not able to run codes in nni 3.0 

I would like to use TensorFlow implementation of DARTs (which is not in nni 2.7). Could you please advise me how I can use it?
